torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+  }
+
+ private:
+  fuseBroadcastWithWarpReduce(const std::vector<kir::Expr*>& exprs) {
+    // open stack space for global scope
+    running_kir_tv_to_allocate_map_.emplace_back(
+        std::make_unique<
+            std::unordered_map<kir::TensorView*, kir::Allocate*>>());
+    running_visible_allocation_stack_.emplace_back(
+        std::make_unique<std::vector<kir::Allocate*>>());
+
+    for (auto expr : exprs) {
+      handle(expr);
+    }
+  }
+
+  void handle(kir::Expr* expr) {
+    if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      handle(for_loop);
+      return;
+    } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+      handle(ite);
+      return;
+    }
+
+    // Process expr inputs if needs replacement
+    for (auto inp : expr->inputs()) {
+      if (auto input_ti = dynamic_cast<kir::TensorIndex*>(inp)) {
+        auto replace = findMaybeReplacedTensorIndex(input_ti);
+        if (replace.has_value()) {
+          val_replacement_map_[input_ti] = replace.value();
+        }
+      }
+    }
+
+    // Handle reduction definitions
+    if (auto reduction = dynamic_cast<kir::ReductionOp*>(expr)) {
+      handle(reduction);
+    } else if (auto broadcast = dynamic_cast<kir::BroadcastOp*>(expr)) {
+      handle(broadcast);
+    } else if (auto allocate = dynamic_cast<kir::Allocate*>(expr)) {
+      handle(allocate);
+    }
+  }
+
+  bool openLoopNestLevel(kir::IterDomain* id) {
+    if (id->isThread()) {
+      return false;
+    }
+    if (id->parallelType() == ParallelType::Serial && id->isBroadcast()) {
+      return false;
+    }
+    return true;
+  }
+
+  void handle(kir::ForLoop* for_loop) {
+    // Keep track of visible reduction outputs
+    bool open_nest_level = openLoopNestLevel(for_loop->iter_domain());
+    if (open_nest_level) {
+      running_kir_tv_to_allocate_map_.emplace_back(
+          std::make_unique<
+              std::unordered_map<kir::TensorView*, kir::Allocate*>>());
+      running_visible_allocation_stack_.emplace_back(
+          std::make_unique<std::vector<kir::Allocate*>>());
+    }
+    for (auto expr : for_loop->body().exprs()) {
+      handle(expr);
+    }
+    if (open_nest_level) {
+      running_kir_tv_to_allocate_map_.pop_back();
+    }
+  }
+
+  void handle(kir::IfThenElse* ite) {
+    running_visible_allocation_stack_.emplace_back(
+        std::make_unique<std::vector<kir::Allocate*>>());
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    running_visible_allocation_stack_.pop_back();
+    running_visible_allocation_stack_.emplace_back(
+        std::make_unique<std::vector<kir::Allocate*>>());
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+    running_visible_allocation_stack_.pop_back();
+  }
+
+  //! Place this allocate on the list of currently visible allocates,
+  //!  organized by loop nest level.
+  void handle(kir::Allocate* allocate) {
```

Comment:
Can you please add some comments why only certain `Allocate` nodes are added to the visible node stack?